### PR TITLE
Introduce generic solver

### PR DESCRIPTION
**Pull request by marusak**

### CODE REVIEW

File: src/domains/predabs_domain.cpp [hunk 1]
```diff
@@ -45,6 +45,84 @@ void predabs_domaint::initialize(valuet &value)
   }
 }
 
+const exprt predabs_domaint::initialize_solver(
+  const local_SSAt &SSA,
+  const exprt &precondition,
+  template_generator_baset &template_generator)
+{
+       get_row_set(todo_preds);
+}
```
Comment (viktormalik) [inline src/domains/predabs_domain.cpp]: The method should return something.

**Comment by marusak** [comment]
@viktormalik updated

File: src/domains/lexlinrank_domain.h [hunk 2]
```diff
@@ -15,6 +15,7 @@ Author: Peter Schrammel
 
 #include <util/std_expr.h>
 #include <util/arith_tools.h>
+#include "../domains/incremental_solver.h"
```
Comment (viktormalik) [inline src/domains/lexlinrank_domain.h]: Use `#include <...>` (multiple occurrences).

File: src/domains/linrank_domain.h [hunk 3]
```diff
@@ -16,6 +16,7 @@ Author: Peter Schrammel
 #include <util/std_expr.h>
 #include <util/arith_tools.h>
 #include <util/ieee_float.h>
+#include "../domains/incremental_solver.h"
```
Comment (viktormalik) [inline src/domains/linrank_domain.h]: Same as above (include).

File: src/domains/ssa_analyzer.cpp [hunk 4]
```diff
@@ -151,41 +157,51 @@ void ssa_analyzert::operator()(
...
-      strategy_solver=new strategy_solver_enumerationt(
-        *static_cast<tpolyhedra_domaint *>(domain), solver, SSA.ns);
+    s_solver=new strategy_solvert(
+      *static_cast<tpolyhedra_domaint *>(domain),
+      solver,
+      SSA,
+      precondition,
+      get_message_handler(),
+      template_generator);
+    result=new tpolyhedra_domaint::templ_valuet();
```
Comment (viktormalik) [inline src/domains/ssa_analyzer.cpp]: Strange indentation of the block.

**Comment by viktormalik** [comment]
@marusak There are errors reported by the linter, please fix them.
Also, re-check indentation everywhere, indent width should be 2 (now it is 4 or even 0 somewhere).

File: src/domains/heap_domain.cpp [hunk 5]
```diff
@@ -1655,3 +1691,354 @@ void heap_domaint::undo_restriction()
...
+  Inputs:
+
+ Outputs:
+
+ Purpose: Find symbolic path that is explored by the current solver
+          iteration. A path is specified by a conjunction of literals
+          containing loop-select guards of all loops in program.
+
+\*******************************************************************/
+void heap_domaint::find_symbolic_path(
```
Comment (viktormalik) [inline src/domains/heap_domain.cpp]: Move this function to `strategy_solver_base`. It willl only accept the `current_guard` parameter, which can be retrieved from the domain (introduce new method for the domain interface: `get_current_loop_guard`).

File: src/domains/strategy_solver.cpp [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+          std::vector<exprt> required_values=
+            domain.get_required_smt_values(row);
+          std::vector<exprt> got_values;
+          for(auto &c_exprt : required_values)
+          {
+            got_values.push_back(solver.solver->get(c_exprt));
+          }
+          domain.set_smt_values(got_values);
+
+          improved=domain.edit_row(row, inv, improved);
```
Comment (viktormalik) [inline src/domains/strategy_solver.cpp]: Add here:
    
    auto current_loop_guard = doimain.get_current_loop_guard();
    find_symbolic_path(current_loop_guard);

File: src/domains/Makefile [hunk 7]
```diff
@@ -3,16 +3,15 @@ SRC = tpolyhedra_domain.cpp equality_domain.cpp domain.cpp \
   heap_tpolyhedra_domain.cpp heap_tpolyhedra_sympath_domain.cpp \
   symbolic_path.cpp\
   ssa_analyzer.cpp util.cpp incremental_solver.cpp \
-  strategy_solver_base.cpp strategy_solver_equality.cpp \
-  linrank_domain.cpp lexlinrank_domain.cpp\
-  ranking_solver_enumeration.cpp lexlinrank_solver_enumeration.cpp \
-  strategy_solver_enumeration.cpp strategy_solver_binsearch.cpp \
+  strategy_solver_binsearch.cpp \
+  strategy_solver_base.cpp \
+  linrank_domain.cpp lexlinrank_domain.cpp \
```
Comment (peterschrammel) [inline src/domains/Makefile]: Put each file on a separate line

File: src/domains/domain.h [hunk 8]
```diff
@@ -54,6 +60,11 @@ class domaint
 
   typedef std::vector<var_spect> var_specst;
 
+  // handles on values to retrieve from model
```
Comment (peterschrammel) [inline src/domains/domain.h]: Why is this required as part of the domain interface?

File: src/domains/Makefile (hunk 7 above)
Comment (marusak) [reply src/domains/Makefile]: This is not really change in this PR. The same line exists upstream (see line 7), I just added there space before backslash (by mistake).

**Comment by marusak** [comment]
@viktormalik Yesterday I pushed fixes for both issues you pointed out. Both are separate commits as squashing of the first one would be rather annoying :) But I am fine with doing it if it is a hard requirement.

File: src/domains/domain.h (hunk 8 above)
Comment (marusak) [reply src/domains/domain.h]: Mostly because all functionality that uses it moved into domain and keeping it in solver would just mean passing references there and back. (and possibly also some better reason that I cannot now remember why I dropped it here)

**Comment by peterschrammel** [comment]
Also, CI is failing...

**Comment by marusak** [comment]
Thanks for pointing this out @peterschrammel , one surplus space :) Now it is green
